ranger/core/actions.py

```diff
@@ -303,91 +321,128 @@ def get_macros(self):  # pylint: disable=too-many-branches,too-many-statements
+                       for fl in self.fm.thistab.get_selection()]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['p'] = get_path
 
-        if self.fm.copy_buffer:
-            macros['c'] = [fl.path for fl in self.fm.copy_buffer]
-        else:
-            macros['c'] = MACRO_FAIL
+        def get_selection():
+            if self.fm.thistab and self.fm.thistab.get_selection:
+                res = [fl.relative_path for fl in self.fm.thistab.get_selection()]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['s'] = get_selection
 
-        if self.fm.thisdir.files:
-            macros['t'] = [fl.relative_path for fl in self.fm.thisdir.files
-                           if fl.realpath in self.fm.tags or []]
-        else:
-            macros['t'] = MACRO_FAIL
+        macros['c'] = lambda: [
+            fl.path for fl in self.fm.copy_buffer] if self.fm.copy_buffer else MACRO_FAIL
 
-        if self.fm.thisdir:
-            macros['d'] = self.fm.thisdir.path
-        else:
-            macros['d'] = '.'
+        def get_tagged():
+            if self.fm.thisdir and self.fm.thisdir.files:
+                res = [fl.relative_path for fl in self.fm.thisdir.files
+                       if fl.realpath in self.fm.tags or []]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['t'] = get_tagged
+
+        macros['d'] = lambda: self.fm.thisdir.path if self.fm.thisdir else '.'
 
         # define d/f/p/s macros for each tab
+        tabdir = None
         for i in range(1, 10):
             try:
                 tab = self.fm.tabs[i]
             except KeyError:
                 continue
+            if not tab:
+                continue
+
             tabdir = tab.thisdir
             if not tabdir:
                 continue
             i = str(i)
-            macros[i + 'd'] = tabdir.path
-            if tabdir.get_selection():
-                macros[i + 'p'] = [os.path.join(tabdir.path, fl.relative_path)
-                                   for fl in tabdir.get_selection()]
-                macros[i + 's'] = [fl.path for fl in tabdir.get_selection()]
-            else:
-                macros[i + 'p'] = MACRO_FAIL
-                macros[i + 's'] = MACRO_FAIL
-            if tabdir.pointed_obj:
-                macros[i + 'f'] = tabdir.pointed_obj.path
-            else:
-                macros[i + 'f'] = MACRO_FAIL
+            macros[i + 'd'] = lambda: tabdir.path
+
+            macros[i + 'p'] = lambda: ([os.path.join(tabdir.path, fl.relative_path)
+                                        for fl in tabdir.get_selection()]
+                                       if tabdir.get_selection() else MACRO_FAIL)
+
+            macros[i + 's'] = lambda: ([fl.path for fl in tabdir.get_selection()]
+                                       if tabdir.get_selection() else MACRO_FAIL)
+
+            macros[i
```

Comment:
Weird split here, please split inside the parens instead.

Reply:
Yep. Probably of that one utility I use for automatic PEP8. Oh, well